platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/schedulers/SequentialThreadScheduler.java

```diff
@@ -0,0 +1,200 @@
+            @NonNull final FractionalTimer busyTimer,
+            @NonNull final Duration sleepDuration,
+            final boolean flushEnabled,
+            final boolean insertionIsBlocking) {
+        super(model, name, TaskSchedulerType.SEQUENTIAL_THREAD, flushEnabled, insertionIsBlocking);
+
+        this.uncaughtExceptionHandler = Objects.requireNonNull(uncaughtExceptionHandler);
+        this.onRamp = Objects.requireNonNull(onRamp);
+        this.offRamp = Objects.requireNonNull(offRamp);
+        this.busyTimer = Objects.requireNonNull(busyTimer);
+        this.sleepDuration = Objects.requireNonNull(sleepDuration);
+
+        thread = new Thread(this::run, "<scheduler " + name + ">");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getUnprocessedTaskCount() {
+        return onRamp.getCount();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void flush() {
+        throwIfFlushDisabled();
+        onRamp.forceOnRamp();
+        final Semaphore semaphore = new Semaphore(1);
+        semaphore.acquireUninterruptibly();
+        tasks.add(new SequentialThreadTask(x -> semaphore.release(), semaphore));
+        semaphore.acquireUninterruptibly();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void put(@NonNull final Consumer<Object> handler, @NonNull final Object data) {
+        onRamp.onRamp();
+        tasks.add(new SequentialThreadTask(handler, data));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean offer(@NonNull final Consumer<Object> handler, @NonNull final Object data) {
+        final boolean accepted = onRamp.attemptOnRamp();
+        if (!accepted) {
+            return false;
+        }
+
+        tasks.add(new SequentialThreadTask(handler, data));
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void inject(@NonNull final Consumer<Object> handler, @NonNull final Object data) {
+        onRamp.forceOnRamp();
+        tasks.add(new SequentialThreadTask(handler, data));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start() {
+        thread.start();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void stop() {
+        alive.set(false);
+    }
+
+    /**
+     * Take work off of the queue and handle it.
+     */
+    private void run() {
+        final List<SequentialThreadTask> buffer = new ArrayList<>(BUFFER_SIZE);
+
+        while (alive.get()) {
+            tasks.drainTo(buffer, BUFFER_SIZE);
+            if (buffer.isEmpty()) {
+                if (sleepDuration.toNanos() > 0) {
+                    try {
+                        NANOSECONDS.sleep(sleepDuration.toNanos());
+                    } catch (final InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        return;
+                    }
```

Comment:
why not do a blocking take? this will cause a sleep/wakeup very often. Some threads are usually idle, like the signed state file writer.

Reply:
Good suggestion, done